modules/library/data/src/main/java/org/geotools/data/memory/MemoryFeatureReader.java

```diff
@@ -30,20 +34,27 @@
 import org.opengis.feature.simple.SimpleFeatureType;
 
 public class MemoryFeatureReader implements FeatureReader<SimpleFeatureType, SimpleFeature>{
-    ContentState state;
+
     SimpleFeatureType featureType;
     Iterator<SimpleFeature> iterator;
 
     public MemoryFeatureReader(ContentState state, Query query) throws IOException {
-        this.state = state;
         featureType = state.getFeatureType();
-        String typeName = getFeatureType().getTypeName();
-        MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
-        iterator = store.features(typeName).values().iterator();
+
+        final MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
+        final Map<String, SimpleFeature> features = store.features(featureType.getTypeName());
+        synchronized (features) {
```

Comment:
Hi Frank, yes, one has to call write() to actually apply the changes.
About the UI notifications, I have no clue, GeoServer has no use of listeners so I'm not up to speed with their implementation and usage. I believe the users for events are using uDig and the gt-swing module, which is presently unmaintained (or, and possibly gt-swt too). Maybe ask on the developer list.